<!DOCTYPE html>



  


<html class="theme-next gemini use-motion" lang="zh-Hans">
<head><meta name="generator" content="Hexo 3.9.0">
  <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform">
<meta http-equiv="Cache-Control" content="no-siteapp">
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css">







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css">

<link href="/css/main.css?v=5.1.4" rel="stylesheet" type="text/css">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png?v=5.1.4">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png?v=5.1.4">


  <link rel="mask-icon" href="/images/logo.svg?v=5.1.4" color="#222">





  <meta name="keywords" content="Hexo, NexT">





  <link rel="alternate" href="/atom.xml" title="生信拾光" type="application/atom+xml">






<meta name="description" content="Seize the day! Collect the moment! Share the memory!">
<meta property="og:type" content="website">
<meta property="og:title" content="生信拾光">
<meta property="og:url" content="http://yoursite.com/index.html">
<meta property="og:site_name" content="生信拾光">
<meta property="og:description" content="Seize the day! Collect the moment! Share the memory!">
<meta property="og:locale" content="zh-Hans">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="生信拾光">
<meta name="twitter:description" content="Seize the day! Collect the moment! Share the memory!">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Gemini',
    version: '5.1.4',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: '六六'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://yoursite.com/">





  <title>生信拾光</title>
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="zh-Hans">

  
  
    
  

  <div class="container sidebar-position-left 
  page-home">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">生信拾光</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br>
            
            首页
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br>
            
            标签
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br>
            
            分类
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br>
            
            归档
          </a>
        </li>
      

      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br>
            
            搜索
          </a>
        </li>
      
    </ul>
  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off" placeholder="搜索..." spellcheck="false" type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            
  <section id="posts" class="posts-expand">
    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/08/23/test/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="六六">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="生信拾光">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/08/23/test/" itemprop="url">test</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-23T11:17:08+08:00">
                2019-08-23
              </time>
            

            

            
          </span>

          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  0
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  1
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            
          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/08/23/第2篇：原始数据的质控、比对和过滤/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="六六">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="生信拾光">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/08/23/第2篇：原始数据的质控、比对和过滤/" itemprop="url">第2篇：原始数据的质控、比对和过滤</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-23T11:17:08+08:00">
                2019-08-23
              </time>
            

            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/组学/" itemprop="url" rel="index">
                    <span itemprop="name">组学</span>
                  </a>
                </span>

                
                
                  ， 
                
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/组学/表观组/" itemprop="url" rel="index">
                    <span itemprop="name">表观组</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  945
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  3
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          <p>这部分内容包括对原始测序数据质控，然后比对过滤，这是所有NGS数据处理的上游分析。</p>
<ul>
<li><strong>ATAC-Seq与其他方法不同的一点是需要过滤去除线粒体（如果是植物，还需要过滤叶绿体），因为线粒体DNA是裸露的，也可以被Tn5酶识别切割。</strong></li>
<li><strong>另外一点需要注意的是课程中给出的是单端比对的示例代码，如果是双端测序做相应更改即可。</strong></li>
</ul>
          <!--noindex-->
          <div class="post-button text-center">
            <a class="btn" href="/2019/08/23/第2篇：原始数据的质控、比对和过滤/#more" rel="contents">
              阅读全文 &raquo;
            </a>
          </div>
          <!--/noindex-->
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/08/23/使用BETA整合分析ATAC-seq:ChIP-seq和RNA-seq数据/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="六六">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="生信拾光">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/08/23/使用BETA整合分析ATAC-seq:ChIP-seq和RNA-seq数据/" itemprop="url">未命名</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-23T11:11:47+08:00">
                2019-08-23
              </time>
            

            

            
          </span>

          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  892
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  3
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h1 id="使用BETA整合分析ATAC-seq-ChIP-seq和RNA-seq数据"><a href="#使用BETA整合分析ATAC-seq-ChIP-seq和RNA-seq数据" class="headerlink" title="使用BETA整合分析ATAC-seq/ChIP-seq和RNA-seq数据"></a>使用BETA整合分析ATAC-seq/ChIP-seq和RNA-seq数据</h1><p><strong><a href="http://cistrome.dfci.harvard.edu/BETA/tutorial.html" target="_blank" rel="noopener">BETA 网址</a></strong>: <a href="http://cistrome.dfci.harvard.edu/BETA/tutorial.html" target="_blank" rel="noopener">http://cistrome.dfci.harvard.edu/BETA/tutorial.html</a></p>
<h3 id="功能介绍"><a href="#功能介绍" class="headerlink" title="功能介绍"></a>功能介绍</h3><p><img src="https://upload-images.jianshu.io/upload_images/8242255-df9dfd4f7834cee4.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt></p>
<h3 id="数据准备："><a href="#数据准备：" class="headerlink" title="数据准备："></a>数据准备：</h3><ul>
<li><p>peak bed file</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"># chr str end</span><br><span class="line">$head mep_ebDfpeak_deseq2_sig.bed</span><br><span class="line">chr8    12050808        12052743</span><br><span class="line">chr9    66492999        66494729</span><br><span class="line">chr4    125632449       125634758</span><br><span class="line">chr4    10182030        10183501</span><br></pre></td></tr></table></figure>
</li>
<li><p>df expression file</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"># symbol log2FC pvalue</span><br><span class="line">$head mep_eb_dfexpre</span><br><span class="line">MELK    6.23721078699297        5.47784458574045e-218</span><br><span class="line">RAD51AP1        4.17541170218602        2.86774188064759e-217</span><br><span class="line">HMGN5   5.98876520905166        7.49359285750074e-216</span><br><span class="line">ZWILCH  3.74468061674801        2.42032701371558e-204</span><br><span class="line">HNRNPA2B1       2.76252025379004        3.05116228100538e-190</span><br><span class="line">HLA-DRA -5.19413815362273       2.27680965999626e-188</span><br></pre></td></tr></table></figure>
</li>
</ul>
<h3 id="软件下载安装"><a href="#软件下载安装" class="headerlink" title="软件下载安装"></a>软件下载安装</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">wget http://cistrome.dfci.harvard.edu/BETA/src/BETA_1.0.7.zip</span><br><span class="line">unzip BETA_1.0.7.zip</span><br><span class="line">cd BETA_1.0.7/</span><br><span class="line">sudo python setup.py install</span><br></pre></td></tr></table></figure>
<h3 id="使用"><a href="#使用" class="headerlink" title="使用"></a>使用</h3><h5 id="BETA-Basic"><a href="#BETA-Basic" class="headerlink" title="BETA Basic"></a>BETA Basic</h5><p>BETA Basic 预测调控因子的功能（激活或抑制）和靶基因<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">BETA basic –p 3656_peaks.bed –e AR_expr.xls –k LIM –g hg19 --da500 –n basic</span><br></pre></td></tr></table></figure></p>
<p><strong>参数：</strong><br><code>-p</code> ： peak bed file (必需参数）<br><code>-e</code>：df expression file (必需参数）<br><code>-k</code>：表达量文件格式是limma的标准格式（根据实际情况修改）<br><code>-g</code>：基因组文件，如hg19,hg38,mm10,mm9<br><code>-n</code>：输出文件前缀，BETA默认是<strong>NA</strong><br><code>-da</code>：选择显著性变化（上调和下调的）表达数据（根据实际情况修改）</p>
<p><strong>其他格式(不是limma和cuffdif标准格式）的差异表达文件格式：</strong></p>
<ul>
<li>需设置参数<code>--info</code>, geneID如果是smbol需要设置<code>--gname2</code></li>
</ul>
<h5 id="BETA-Plus"><a href="#BETA-Plus" class="headerlink" title="BETA Plus"></a>BETA Plus</h5><p>BETA Plus 预测调控因子的功能（激活或抑制）和靶基因，以及靶向区域的motif分析</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">BETA plus –p 3656_peaks.bed –e AR_expr.xls –k LIM –g hg19 --gs hg19.fa --bl</span><br></pre></td></tr></table></figure>
<p><strong>参数：</strong></p>
<p><code>--gs</code> : motif分析需要的参数, 基因组fasta格式<br><code>--bl</code> :可选参数，当考虑boundary（如CTCF）时需要的参数</p>
<h4 id="BETA-Minus"><a href="#BETA-Minus" class="headerlink" title="BETA Minus"></a>BETA Minus</h4><p>BETA Minus 只根据peak bed数据基于调控潜能值预测靶基因<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">BETA minus -p 3656_peaks.bed --bl -g hg19</span><br></pre></td></tr></table></figure></p>
<h5 id="其他可选参数"><a href="#其他可选参数" class="headerlink" title="其他可选参数"></a>其他可选参数</h5><p><code>-n</code>：结果文件的名字<br><code>-o</code>:输出文件路径<br><code>--gname2</code>：geneID如果是smbol需要设置<code>--gname2</code><br><code>--info</code>：其他格式(不是limma和cuffdif标准格式）的差异表达文件格式：需设置参数<code>--info</code>,<br><code>--pn</code>:peak 数目，默认是10，000<br><code>-d DISTANCE,--distance</code> 距离TSS的距离. DEFAULT=100000 (100kb)<br><code>--df DIFF_FDR</code> : 输入0-1的数值，表示统计的阈值，如FDR等，默认是1.<br><code>--da DIFF_AMOUNT</code>; 取显著表达的基因，0-1表示比例，大于1表示个数。如2000，表示去top 2000个上调和下调的差异基因；0.5表示取50%上调和下调的基因，默认是0.5。<br><code>-c CUTOFF, --cutoff</code>： 0~1的一个阈值。默认是1e-3。<br><code>-r REFERENCE, --reference</code>： 当参考基因组不是hg19,或mm9时使用此参数。<br><code>--bl BOUNDARY</code> :LIMIT Boolean Value. Whether or not use CTCF boundary to get a peak’s associated gene, DEFAULT=FALSE<br><code>-bf BOUNDARYFILE</code> : Some BED format boundary file, use this only when You set –bl and the genome is neither hg19 nor mm9</p>
<h3 id="相关资料"><a href="#相关资料" class="headerlink" title="相关资料"></a>相关资料</h3><p><strong>文章</strong>Wang, S., Sun, H., Ma, J., Zang, C., Wang, C., Wang, J., … &amp; Liu, X. S. (2013). Target analysis by integration of transcriptome and ChIP-seq data with BETA. Nature protocols, 8(12), 2502-2515.<br><strong>PMID:</strong> 24263090<br><strong>网址：</strong> <a href="http://cistrome.org/BETA/index.html" target="_blank" rel="noopener">http://cistrome.org/BETA/index.html</a></p>
<h3 id="使用代码"><a href="#使用代码" class="headerlink" title="使用代码"></a>使用代码</h3><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">/software/biosoft/software/python/python2.7/bin/BETA plus -P mep_ebDfpeak_deseq2_sig.bed --info -e mep_eb_dfexpre --gname2 -g hg 19 --gs hg19.fa --bl</span><br></pre></td></tr></table></figure>
<h3 id="几点疑问"><a href="#几点疑问" class="headerlink" title="几点疑问"></a>几点疑问</h3><p>BETA最初的设计主要是用于ChIP-Seq和RNA-seq的整合分析，最终可以预测特定一个转录因子的靶基因，以及这些基因是激活还是抑制的功能。但是对于ATAC-seq，它得到的信号整个基因组范围转录因子的信号，那么与转录组整合分析得到的结果怎么解释？只能知道一些基因的功能是激活的或抑制的，但是并不能知道这些基因是哪些转录因子的靶基因。所以BETA是否适合ATAC-seq和与RNA-seq的整合分析？如果可以，整合分析的结果能得到哪些有价值的信息？</p>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/08/23/第5篇：对ATAC-SeqChIP-seq的质量评估（二）——ChIPQC/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="六六">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="生信拾光">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/08/23/第5篇：对ATAC-SeqChIP-seq的质量评估（二）——ChIPQC/" itemprop="url">未命名</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-23T11:11:47+08:00">
                2019-08-23
              </time>
            

            

            
          </span>

          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1.8k
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  7
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h1 id="第5篇：对ATAC-Seq-ChIP-seq的质量评估（二）——ChIPQC"><a href="#第5篇：对ATAC-Seq-ChIP-seq的质量评估（二）——ChIPQC" class="headerlink" title="第5篇：对ATAC-Seq/ChIP-seq的质量评估（二）——ChIPQC"></a>第5篇：对ATAC-Seq/ChIP-seq的质量评估（二）——ChIPQC</h1><h3 id="1-学习目标"><a href="#1-学习目标" class="headerlink" title="1. 学习目标"></a>1. 学习目标</h3><ul>
<li>讨论ChIP-seq数据质量评估的其他方法</li>
<li>用ChIPQC产生质量统计报告</li>
<li>鉴定低质量数据的来源<br><strong>概览图</strong><br><img src="https://upload-images.jianshu.io/upload_images/8242255-df6c67811efb85eb.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Additional Quality Metrics for ChIP-seq data"><br>ENCODE评估数据质量采用多种指标，如前面已经讨论过的链相关的指标NSC和RSC。这一节将会讨论评估信号分布的其他指标。<blockquote>
<p><strong>NOTE</strong>：这里给出的评估指标只是反映数据质量的好坏，符合阈值的并不意味着实验是成功的，不符合阈值的也不一定意味着失败。</p>
</blockquote>
<h3 id="2-常见质量评估指标的介绍"><a href="#2-常见质量评估指标的介绍" class="headerlink" title="2.常见质量评估指标的介绍"></a>2.常见质量评估指标的介绍</h3></li>
<li><h4 id="SSD"><a href="#SSD" class="headerlink" title="SSD"></a>SSD</h4>SSD值是对富集效果的评估。SSD值依赖于全基因组的pile-up信号强度，对真实的ChIP富集和干扰的强信号区域都很敏感。SSD值越大表明富集越好。<blockquote>
<p>“It provides a measure of pileup across the genome and is computed by looking at the standard deviation of signal pile-up along the genome normalised to the total number of reads. ”</p>
</blockquote>
</li>
<li><h4 id="FRiP-Fraction-of-reads-in-peaks"><a href="#FRiP-Fraction-of-reads-in-peaks" class="headerlink" title="FRiP:Fraction of reads in peaks"></a>FRiP:Fraction of reads in peaks</h4>FRiP表示的是peaks中的reads与总reads的比例。它是另一个反映样本富集效果或IP好坏的评价指标。可以理解为是“信噪比”即文库中结合位点片段占背景reads的比例。<strong>一个典型质量好的TF富集FRiP值约5%或者更高，polII的FRiP值约为30%或者更高，也有一些质量好的数据FRiP值&lt;1%（如RNAPIII）</strong></li>
<li><h4 id="Relative-Enrichment-of-Genomic-Intervals-REGI"><a href="#Relative-Enrichment-of-Genomic-Intervals-REGI" class="headerlink" title="Relative Enrichment of Genomic Intervals (REGI)"></a>Relative Enrichment of Genomic Intervals (REGI)</h4>REGI是对peaks在不同基因组特征位点分布的统计。</li>
<li><h4 id="RiBL-Reads-overlapping-in-Blacklisted-Regions"><a href="#RiBL-Reads-overlapping-in-Blacklisted-Regions" class="headerlink" title="RiBL: Reads overlapping in Blacklisted Regions"></a>RiBL: Reads overlapping in Blacklisted Regions</h4>过滤人工造成的高信号区域非常重要，如ENCIDE和modENCODE提供的DAC Blacklisted Regions track。这些区域经常在特定的重复序列处出现，如着丝粒、端粒、卫星重复序列等，通过简单的比对过滤是不能去除的。来自blacklisted regions的信号会造成call peak 和片段长度评估的混淆。<br>RiBL值可以表示背景信号或input的信号水平，与input sample的SSD值以及input和ChIP sample的读长覆盖值相关。这些区域通常是基因组的0.5%，或者更高的比例（10%）。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-ffa899faf5d8e033.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt></li>
</ul>
<h3 id="3-ChIPQC-quality-metrics-report"><a href="#3-ChIPQC-quality-metrics-report" class="headerlink" title="3. ChIPQC: quality metrics report"></a>3. ChIPQC: quality metrics report</h3><p>ChIPQC是一个Bioconductor包，输入文件包括BAM和peak文件，可以自动计算一些质量评估值，并产生质量报告。</p>
<h4 id="准备数据"><a href="#准备数据" class="headerlink" title="准备数据"></a>准备数据</h4><ul>
<li><strong>BAM files</strong><br>首先对比对过滤后的bam数据（chr12_aln.bam）建索引，然后将bam和index文件从<code>~/ngs_course/chipseq/results/bowtie2</code>移动到自己的目录文件夹<code>data/bams</code></li>
<li><strong>peak files</strong><br>将narrowPeak 文件从macs2目录下<code>~/ngs_course/chipseq/results/macs2</code> 移动到自己目录下<code>data/peakcalls</code></li>
<li><p><strong>sampleSheet file</strong><br>sampleSheet file是唯一需要自己根据实验设计和数据存储地址等信息创建的一个csv格式文件（bam,peak文件分别在比对和call peak的步骤产生）。sampleSheet具体需要包含的信息如下：<br><img src="https://upload-images.jianshu.io/upload_images/8242255-3910ab7fab88b880.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="sampleSheet header"></p>
</li>
<li><p>SampleID: 样本ID</p>
</li>
<li>Tissue, Factor, Condition: 不同的实验设计对照信息，三列信息必须包含在sampleSheet里，如果没有某一列的信息设为NA。</li>
<li>Replicate : 重复样本的编号</li>
<li>bamReads : 实验组BAM 文件的路径（data/bams）</li>
<li>ControlID : 对照组样本ID</li>
<li>bamControl :对照组样本的bam文件路径</li>
<li>Peaks :样本peaks文件的路径</li>
<li>PeakCaller ：peak类型的字符串，可以是raw,bed,narrow,macs等。<h4 id="下载安装ChIPQC"><a href="#下载安装ChIPQC" class="headerlink" title="下载安装ChIPQC"></a>下载安装ChIPQC</h4><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">source(&quot;http://bioconductor.org/biocLite.R&quot;)</span><br><span class="line">biocLite(&quot;ChIPQC&quot;)</span><br></pre></td></tr></table></figure>
</li>
</ul>
<h4 id="Running-ChIPQC"><a href="#Running-ChIPQC" class="headerlink" title="Running ChIPQC"></a>Running ChIPQC</h4><p>ChIPQC只需要三步就可以完成质量评估和报告生成。</p>
<ul>
<li><p><strong>首先载入包和sampleSheet信息</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">## Load libraries</span><br><span class="line">library(ChIPQC)</span><br><span class="line">## Load sample data</span><br><span class="line">samples &lt;- read.csv(&apos;meta/samplesheet_chr12.csv&apos;)</span><br><span class="line">View(samples)</span><br></pre></td></tr></table></figure>
</li>
<li><p><strong>创建ChIPQC对象</strong><br>利用sampleSheet的信息读取每个样本的bam和narrowpeak文件，并计算质量评估值，结果存在一个对象里。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">## Create ChIPQC object</span><br><span class="line">chipObj &lt;- ChIPQC(samples, annotation=&quot;hg19&quot;)</span><br></pre></td></tr></table></figure>
</li>
<li><p><strong>生成ChIPQC报告</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">## Create ChIPQC report</span><br><span class="line">ChIPQCreport(chipObj, reportName=&quot;ChIP QC report: Nanog and Pou5f1&quot;, reportFolder=&quot;ChIPQCreport&quot;)</span><br></pre></td></tr></table></figure>
</li>
</ul>
<h4 id="ChIPQC报告解读"><a href="#ChIPQC报告解读" class="headerlink" title="ChIPQC报告解读"></a>ChIPQC报告解读</h4><p>ChIPQC生成的结果包含一个网页报告和报告中含有的所有图片。<br>网页报告有三部分：<strong>QC Summary ；QC Results；QC files and versions</strong><br><img src="https://upload-images.jianshu.io/upload_images/8242255-a929302bfd47dc9d.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt></p>
<h5 id="（1）QC-Summary-Overview-of-results"><a href="#（1）QC-Summary-Overview-of-results" class="headerlink" title="（1）QC Summary - Overview of results"></a>（1）QC Summary - Overview of results</h5><p><img src="https://upload-images.jianshu.io/upload_images/8242255-01749f12cf83c9af.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Table 1"><br>QC summary包含sampleSheet里填写的样本的基本信息Tissue,Factor,Condition,Replicate。另外还有上面提到的质量评估的常用指标SSD、RiP%和RiBL值。越高的SSD值表明富集效果越好，Pou5f1样本（2.6,3）有较高的SSD值，RiBL值不是很高，FRiP的比例在5%附近或者更高，除了Pou5f1-rep2。</p>
<ul>
<li><strong>SSD</strong> - SSD score (htSeqTools)</li>
<li><strong>RIP%</strong> - Percentage of reads wthin peaks</li>
<li><strong>RIBL%</strong> - Percentage of reads wthin Blacklist regions<br>同时表格中还给出了其他统计信息：</li>
<li><strong>Reads</strong> - Number of sample reads within analysed chromosomes.</li>
<li><strong>Dup%</strong> - Percentage of MapQ filter passing reads marked as duplicates</li>
<li><strong>FragLen</strong> - Estimated fragment length by cross-coverage method</li>
<li><strong>FragLenCC</strong> - Cross-Coverage score at the fragment length</li>
<li><strong>RelativeCC</strong> - Cross-coverage score at the fragment length over Cross-coverage at the read length</li>
</ul>
<h5 id="（2）QC-Results-Full-QC-results-and-figures"><a href="#（2）QC-Results-Full-QC-results-and-figures" class="headerlink" title="（2）QC Results - Full QC results and figures"></a>（2）QC Results - Full QC results and figures</h5><ul>
<li><strong>Mapping, Filtering and Duplication rate</strong><br>第一部分是比对、过滤和重复率质检结果，包括Table2 、Figure1和Figure2。<br><strong>Table 2</strong>主要给出了比对质量和重复率，因为BAM文件是过滤后的，所以这里Dup%都是0.<br><img src="https://upload-images.jianshu.io/upload_images/8242255-ae05107ee85bb02a.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Table 2"><blockquote>
<ul>
<li><strong>Total Dup%</strong>-Percentage of all mapped reads which are marked as duplicates.</li>
<li><strong>Pass MapQ Filter%</strong>-Percentage of all mapped reads whichpass MapQ quality filter</li>
<li><strong>Pass MapQ Filter and Dup%</strong>-Percentage of all reads which pass MapQ filter and are marked asduplicates.</li>
</ul>
</blockquote>
</li>
</ul>
<p><strong>Figure 1</strong>展现了reads在blacklists中的比例，<br><img src="https://upload-images.jianshu.io/upload_images/8242255-c4274afa8e04dd99.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Figure 1.  Barplot of the percentage of reads in blacklists"></p>
<p><strong>Figure 2</strong>是用基因组注释呈现了reads在基因组特征位置如启动子的分布。这幅图里显示在启动子区域富集最明显。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-bf79c7e6edea3523.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Figure 2 Heatmap of log2 enrichment of reads in genomic features"></p>
<ul>
<li><strong>ChIP signal Distribution and Structure</strong><br>第二部分是ChIP信号分布和结构组成，包括Figure3和4。<br><strong>Figure 3</strong>是一个coverage plot， x轴代表在某bp位置read pileup的高度，y轴代表有多少位置有相同的pileup 高度(取log)。**有好的富集的ChIP样本会有一个tail，即更多的位置（y值大）有较高的测序深度。在我们的数据集中Nanog样本与Pou5f1 相比有较高的tails,尤其是重复样本2。但是Pou5f1有较高的SSD值。当SSD高但是coverage看起来低时，可能是存在大片段深度高的区域出现在blacklist 基因组区域。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-00577ac95d666739.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Figure 3.  Plot of the log2 base pairs of genome at differing read depths"></li>
<li><strong>Peak Profile and ChIP Enrichment</strong><br>第3部分是peak的谱图和ChIP的富集，每个peak都集中在summit位置（summit 理解为peak的最高峰值点处）<br><img src="https://upload-images.jianshu.io/upload_images/8242255-d50435914800593a.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Figure 5.  Plot of the average signal profile across peaks"><br>peak的性状取决于研究对象的类型，如转录因子、组蛋白标记、或其他DNA结合蛋白如聚合酶等，相同类型的对象通常有独特特征的谱图。<br><strong>Figure6和7</strong>都是对比对到peak中的reads统计。富集效果好的ChIP样本的reads与peaks会有高比例的重合。尽管Nanog有较高的RiP,但是两个重复样本间的差异大于Pou5f1。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-fec14f72728ddcd7.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Figure 6.  Barplot of the percentage number of reads in peaks"><br><img src="https://upload-images.jianshu.io/upload_images/8242255-c4871f68579ba322.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Figure 7.  Density plot of the number of reads in peaks"><br><strong>Figure8和9</strong>表示样本的聚类效果，分别是相关性聚类热图和PCA。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-ee0ccb984d0ea9dc.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Figure 8.  Plot of correlation between peaksets"><br><img src="https://upload-images.jianshu.io/upload_images/8242255-4f61fd17225198b9.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Figure 9.  PCA of peaksets"><h3 id="4-实验偏差：ChIP-seq数据质量低的来源"><a href="#4-实验偏差：ChIP-seq数据质量低的来源" class="headerlink" title="4. 实验偏差：ChIP-seq数据质量低的来源"></a>4. 实验偏差：ChIP-seq数据质量低的来源</h3></li>
<li>免疫沉淀的特异性和有效性<br>影响因素如抗体的特异性，结和沉淀的强度</li>
<li>片段化<br>超声裂解产生不同大小的片段可能引入偏差</li>
<li>文库构建时的偏差<br>如PCR扩增<h4 id="参考资料："><a href="#参考资料：" class="headerlink" title="参考资料："></a>参考资料：</h4><a href="https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/tree/master/sessionV/lessons" target="_blank" rel="noopener">哈佛深度NGS数据分析课程</a><br><a href="https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/blob/master/sessionV/lessons/05_QC_quality_metrics.md" target="_blank" rel="noopener">05-ChIP-Seq Quality Assessment</a></li>
</ul>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/08/23/第4篇：对ATAC-SeqChIP-seq的质量评估（一）——phantompeakqualtools/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="六六">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="生信拾光">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/08/23/第4篇：对ATAC-SeqChIP-seq的质量评估（一）——phantompeakqualtools/" itemprop="url">未命名</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-23T11:11:47+08:00">
                2019-08-23
              </time>
            

            

            
          </span>

          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1.6k
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  6
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h1 id="第4篇：对ATAC-Seq-ChIP-seq的质量评估（一）——phantompeakqualtools"><a href="#第4篇：对ATAC-Seq-ChIP-seq的质量评估（一）——phantompeakqualtools" class="headerlink" title="第4篇：对ATAC-Seq/ChIP-seq的质量评估（一）——phantompeakqualtools"></a>第4篇：对ATAC-Seq/ChIP-seq的质量评估（一）——phantompeakqualtools</h1><h3 id="学习目标："><a href="#学习目标：" class="headerlink" title="学习目标："></a>学习目标：</h3><ul>
<li>探讨ChIP-seq数据质量低的来源</li>
<li>理解链交叉相关性（ strand cross-correlation）</li>
<li>使用<code>phantompeakqualtools</code>计算交叉相关性和其他相关的质控度量值</li>
<li>评估交叉相关图<h3 id="ChIP-Seq质量评估"><a href="#ChIP-Seq质量评估" class="headerlink" title="ChIP-Seq质量评估"></a>ChIP-Seq质量评估</h3>在下游分析前，最好是先对peak calling 后的ChIP-Seq数据进行质量评估。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-025dd4ff30ae61b8.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt><h3 id="链交叉相关（Strand-cross-correlation）"><a href="#链交叉相关（Strand-cross-correlation）" class="headerlink" title="链交叉相关（Strand cross-correlation）"></a>链交叉相关（Strand cross-correlation）</h3>链交叉相关是一个有效的评估ChIP-Seq质量的方法，它不依赖于peak calling，而是基于ChIP-Seq实验。如果ChIP-Seq实验成功，DNA富集序列标签（蛋白质相互作用的序列）会在reads的双峰富集中产生显著的聚集。<br><strong>产生reads的双峰富集的原因如下：</strong><br>在ChIP-Seq实验中，DNA被片段化，蛋白质结合的片段会被免疫沉淀，所以产生了有蛋白质结合的DNA片段（fragments ）。<br>DNA的正链从5’端开始被测序（如下图红色reads），DNA负链也从5’末端被测序产生如下图所示的蓝色reads。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-9095a287381f5ec6.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Nat Biotechnol. 2008 Dec; 26(12): 1351–1359"><br>由于从DNA片段的5′末端测序，使+链reads的富集（下图中的蓝色部分）与负链reads的富集（下图红色部分）有少量的相互抵消区域。我们<strong>需要确定峰位移多少碱基数目可以在两个峰间产生最大的相关性。</strong>我们可以用<strong>交叉相关的度量值（cross-correlation metric）</strong>计算产生最大相关的位移。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-120795238a4bbaba.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt></li>
</ul>
<p><strong>交叉相关性度量值</strong><br>交叉相关度量是在Watson移动k个碱基后，计算Crick链与Watson链之间的Pearson线性相关。如下面的示意图：<br>首先在链位移为0时，两个向量之间的Pearson相关值为0.539。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-d2316a73a5af7331.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt><br>在链位移5bp，两个向量之间的Pearson相关值为0.931。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-c0b37466457a516e.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt><br>继续移动这些向量，对于每个链位移计算一个相关值。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-e931dc07c84629d3.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt><br>最后，我们将有一个每个碱基对移位与皮尔森相关值的对应表。这是针对每个染色体的每一个峰计算的，然后该值乘以一个缩放因子，再对所有染色体的值相加，就可以绘交叉相关值（y轴）相对于移位值（x轴）生成的交叉相关图。<br><strong>典型的交叉相关图会产生两个峰</strong>：一个富集峰与主要的<strong>片段长度（predominant fragment length）</strong>相关（高相关性），另一个与<strong>read 长度（read length）</strong>相关，这个峰也被称为虚幻峰（“phantom” peak）。</p>
<ul>
<li>质量好的ChIP-Seq数据集倾向产生一个大的<strong>片段长度峰（fragment-length peak）</strong>，下图展示了一个来自于人细胞CTCF（zinc-finger transcription factor）的强信号。如果有好的抗体，转录因子通常产生45,000~60,000个peaks。下图红色的垂直线表示主峰的真实位移，蓝色的垂直线处有一个小的起伏表示read lenngth。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-f291e6c7288d5df6.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="strong signal"></li>
<li>下图是<strong>weaker signal</strong>的一个示例。这里<strong>Pol2</strong>的抗体不是很有效，有分散的峰。交叉相关图中有两个峰，一个是真实的峰位移（185-200bp），另外一个在read length。信号弱的数据集中read length的峰会成为主峰。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-2846dc6bb5085bf8.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt=" weaker signal"></li>
<li>一个失败的实验产生的交叉相关图类似于<strong>input</strong>，在<strong>fragment length</strong>处很少或没有峰，在<strong>read length</strong>处有信号非常强的。这种现象的原因可能是在结合位点附近<strong>fragments</strong>没有显著聚集。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-8781f9f9bd780b15.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="failed experiment"></li>
</ul>
<h3 id="交叉相关性质量评估度量值"><a href="#交叉相关性质量评估度量值" class="headerlink" title="交叉相关性质量评估度量值"></a>交叉相关性质量评估度量值</h3><p>交叉相关谱图可以计算评估ChIP_Seq实验信噪比的度量值，并且实验设计确保<strong>fragment length</strong>准确。低信噪比和不准确的fragment length 表明ChIP-Seq实验可能有问题。<br><strong>Normalized strand cross-correlation coefficent (NSC)：</strong><br>NSC是最大交叉相关值除以背景交叉相关的比率(所有可能的链转移的最小交叉相关值)。<strong>NSC值越大表明富集效果越好，NSC值低于1.1<br>表明较弱的富集，小于1表示无富集。</strong>NSC值稍微低于1.05，有较低的信噪比或很少的峰，这肯能是生物学真实现象，比如有的因子在特定组织类型中只有很少的结合位点；也可能确实是数据质量差。<br><strong>Relative strand cross-correlation coefficient (RSC):</strong><br>RSC是片段长度相关值减去背景相关值除以phantom-peak相关值减去背景相关值。<strong>RSC的最小值可能是0，表示无信号；富集好的实验RSC值大于1；低于1表示质量低。</strong></p>
<h3 id="phantompeakqualtools"><a href="#phantompeakqualtools" class="headerlink" title="phantompeakqualtools"></a>phantompeakqualtools</h3><p>phantompeakqualtools 是一个用于计算ChIP-Seq数据富集和质量度量值的一个工具包。我们将使用该包来计算基于链交叉相关峰的主要插入大小(fragment length)和基于相对phantom peak的数据质量度量值。<code>phantompeakqualtools</code>是一个R包，依赖<code>samtools</code>。<br><strong>下载<code>phantompeakqualtools</code></strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">wget https://storage.googleapis.com/google-code-archive-downloads/v2/code.google.com/phantompeakqualtools/ccQualityControl.v.1.1.tar.gz</span><br><span class="line">tar -xzf ccQualityControl.v.1.1.tar.gz</span><br><span class="line">cd phantompeakqualtools</span><br><span class="line"># 查看README</span><br><span class="line">less README.txt</span><br></pre></td></tr></table></figure></p>
<p><strong>Linux下安装</strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">R </span><br><span class="line">install.packages(&quot;caTools&quot;, lib=&quot;~/R/library&quot;)</span><br></pre></td></tr></table></figure></p>
<p><strong>运行phantompeakqualtools</strong><br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">mkdir -p logs qual</span><br><span class="line"></span><br><span class="line">for bam in bam_dir/sample1.final.bam bam_dir/sample2.final.bam</span><br><span class="line">do </span><br><span class="line">bam2=`basename $bam .final.bam`</span><br><span class="line">Rscript run_spp_nodups.R -c=$bam -savp -out=qual/$&#123;bam2&#125;.qual &gt; logs/$&#123;bam2&#125;.Rout</span><br><span class="line">done</span><br></pre></td></tr></table></figure></p>
<p><strong>参数含义：</strong></p>
<ul>
<li><code>-c</code>: 比对过滤后的bam文件的全路径和名字</li>
<li><code>-savp</code>:保存交叉相关图</li>
<li><code>-out</code>：会产生数据集重要特征值的输出文件</li>
</ul>
<p><strong>输出文件解读</strong><br>输出文件会产生一个tab分割的名为qual的文件,包含的信息如下：</p>
<ul>
<li>COL1：Filename：比对过滤的bam文件名</li>
<li>COL2：numReads ：有效的测序深度</li>
<li>COL3：estFragLen：逗号分隔的交叉相关峰以相关性递减顺序排列的值</li>
<li>COL4: corr_estFragLen: 逗号分隔的以递减顺序排列交叉相关值</li>
<li>COL5: phantomPeak: Read length/phantom peak链位移</li>
<li>COL6: corr_phantomPeak: phantom peak相关值</li>
<li>COL7: argmin_corr:交叉相关最小的链位移</li>
<li>COL8: min_corr:交叉相关最小值</li>
<li>COL9: Normalized strand cross-correlation coefficient (NSC) = COL4 / COL8</li>
<li>COL10: Relative strand cross-correlation coefficient (RSC) = (COL4 - COL8) / (COL6 - COL8)</li>
<li>COL11: QualityTag: Quality tag based on thresholded RSC (codes: -2:veryLow,-1:Low,0:Medium,1:High,2:veryHigh)<br><strong>我们最关注的值是第9列和第11列。</strong><br><strong>Cross-correlation plots</strong><br>课程中的示例数据Nanog_rep1的交叉相关图<br><img src="https://upload-images.jianshu.io/upload_images/8242255-4e872e4a5d5fcdbb.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt><h3 id="参考资料"><a href="#参考资料" class="headerlink" title="参考资料"></a>参考资料</h3><a href="https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/tree/master/sessionV/lessons" target="_blank" rel="noopener">哈佛深度NGS数据分析课程</a><br><a href="https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/blob/master/sessionV/lessons/04_QC_cross_correlation.md" target="_blank" rel="noopener">04_ChIP-Seq Quality Assessment: Cross-correlation</a></li>
</ul>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/08/23/第10篇：ATAC-Seq、ChIP-Seq、RNA-Seq整合分析/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="六六">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="生信拾光">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/08/23/第10篇：ATAC-Seq、ChIP-Seq、RNA-Seq整合分析/" itemprop="url">未命名</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-23T11:11:47+08:00">
                2019-08-23
              </time>
            

            

            
          </span>

          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  889
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  3
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          
            <h1 id="第10篇：ATAC-Seq、ChIP-Seq、RNA-Seq整合分析"><a href="#第10篇：ATAC-Seq、ChIP-Seq、RNA-Seq整合分析" class="headerlink" title="第10篇：ATAC-Seq、ChIP-Seq、RNA-Seq整合分析"></a>第10篇：ATAC-Seq、ChIP-Seq、RNA-Seq整合分析</h1><p>由于基因表达调控机制的复杂性，多种组学数据的整合分析，从不同的层面探究生物问题越来越重要。从RNA-Seq层面，我们可以探究哪些基因具有显著差异，上调或下调；从ChIP-Seq层面，我们可以研究某个特定转录因子的调控作用；从ATAC-Seq我们可以了解到染色质可及性的动态变化，由于染色质的可及性与调控元件或转录因子的结合密切相关，在转录调控中发挥着重要的作用。因此，整合分析可以进一步探究调控某一生物学过程的关键因子（包括顺式调控元件和转录因子），以及哪个转录因子调控了感兴趣的基因，以及感兴趣的转录因子的靶基因等。<br>目前，好像并没有标准化的方法用来整合比较这三种数据。不过在文献中可以看到有很多相同的或不同的思路和方法做整合分析，大家可以在学习交流群中推荐文献，一起解读学习。<br>下面的内容主要介绍这一节课程中RNA-Seq和ChIP-Seq的整合分析中提到的两种方法：<br><strong>一是直接比较</strong>，即首先得到差异基因与ChIP-Seq靶基因的overlap，然后选择一些关键基因比较一下谱图。<br><img src="https://upload-images.jianshu.io/upload_images/8242255-3a8682b8a3260f0c.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Gao et. al, Nature 2014，doi: 10.1038/nature13921"><br><strong>课程中提到的另一种方法是使用BETA工具：</strong><br><strong>BETA (Binding and Expression Target Analysis)</strong>是 Shirley Liu实验室开发的工具，通过整合转录因子或染色质调控因子的ChIP-Seq与差异基因的表达直接预测靶基因，而且有可能发现增强子区的蛋白质的靶基因。<br>BETA有三个功能：</p>
<ul>
<li>预测转录因子的功能是激活还是抑制</li>
<li>预测转录因子的靶基因</li>
<li>鉴定转录因子的motif以及调控转录因子激活或抑制的其他因子</li>
</ul>
<p>BETA包括三个命令取决于输入数据格式和想要的输出数据（可以参考嘉因的帖子<a href="https://mp.weixin.qq.com/s/Nm47ZMbdlIx-OI5j5wo4nA" target="_blank" rel="noopener">ChIP-seq和RNA-seq整合分析，BETA最擅长 | 自己分析数据的完美解决方案</a>）</p>
<ul>
<li>BETA basic: 预测转录因子的功能是激活还是抑制，直接检测靶标</li>
<li>BETA plus：BETA basic + 靶标区域的motif分析</li>
<li>BETA minus：只基于结合数据的调控潜能的值预测TF靶基因</li>
</ul>
<p><img src="https://upload-images.jianshu.io/upload_images/8242255-126e213bad02247c.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="Nat Protoc. 2013 Dec; 8(12): 2502–2515."></p>
<p><strong>工作原理</strong></p>
<ul>
<li>每个基因的调控潜能值是通过计算基因TSS的指定范围内的所有结合位点来计算的</li>
<li>调控潜能是基因受因子调节的可能性，取决于TSS范围的结合位点数以及与结合位点与TSS之间的距离</li>
<li>BETA minus按照调控潜能值对基因排序来鉴定靶标</li>
<li>BETA basic需要特定格式的差异分析结果和显著性的结果。它使用CDF来判断上调基因和下调基因是否与NON-DE不同，这是用来识别激活和抑制功能。使用调控潜能和差异统计量计算每一个基因的得分排序，该得分排序可以用于识别靶标。<h3 id="参考资料："><a href="#参考资料：" class="headerlink" title="参考资料："></a>参考资料：</h3></li>
</ul>
<ol>
<li><a href="https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/blob/master/sessionV/lessons/integrating_rna-seq_and_chip-seq.md" target="_blank" rel="noopener">哈佛深度NGS数据分析课程，10-Integrating RNA-seq and ChIP-seq</a></li>
<li><a href="https://mp.weixin.qq.com/s/-7Bx29H5nM-LQJiu9hbYjw" target="_blank" rel="noopener">大牛新宠，转录调控研究神器！找关键转录因子、推测转录因子-靶基因关系</a></li>
</ol>

          
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
      

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/08/23/第1篇：ATAC-seq的背景介绍以及与ChIP-Seq的异同/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="六六">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.gif">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="生信拾光">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">
                
                <a class="post-title-link" href="/2019/08/23/第1篇：ATAC-seq的背景介绍以及与ChIP-Seq的异同/" itemprop="url">第1篇：ATAC-seq的背景介绍以及与ChIP-Seq的异同</a></h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">发表于</span>
              
              <time title="创建于" itemprop="dateCreated datePublished" datetime="2019-08-23T08:26:56+08:00">
                2019-08-23
              </time>
            

            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">分类于</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/组学/" itemprop="url" rel="index">
                    <span itemprop="name">组学</span>
                  </a>
                </span>

                
                
                  ， 
                
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/组学/表观组/" itemprop="url" rel="index">
                    <span itemprop="name">表观组</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
          

          
          

          

          
            <div class="post-wordcount">
              
                
                <span class="post-meta-item-icon">
                  <i class="fa fa-file-word-o"></i>
                </span>
                
                  <span class="post-meta-item-text">字数统计&#58;</span>
                
                <span title="字数统计">
                  1.5k
                </span>
              

              
                <span class="post-meta-divider">|</span>
              

              
                <span class="post-meta-item-icon">
                  <i class="fa fa-clock-o"></i>
                </span>
                
                  <span class="post-meta-item-text">阅读时长 &asymp;</span>
                
                <span title="阅读时长">
                  5
                </span>
              
            </div>
          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        
          <h3 id="ATAC-Seq简介"><a href="#ATAC-Seq简介" class="headerlink" title="ATAC-Seq简介"></a>ATAC-Seq简介</h3><p><strong>ATAC-seq（Assay for Transposase-Accessible Chromatin with high throughput sequencing）</strong> 是2013年由斯坦福大学<a href="http://greenleaf.stanford.edu/index.html" target="_blank" rel="noopener">William J. Greenleaf</a>和<a href="http://changlab.stanford.edu/index.html" target="_blank" rel="noopener">Howard Y. Chang</a>实验室开发的用于研究染色质可及性（通常也理解为染色质的开放性）的方法， 原理是通过转座酶Tn5容易结合在开放染色质的特性，然后对Tn5酶捕获到的DNA序列进行测序。</p>
          <!--noindex-->
          <div class="post-button text-center">
            <a class="btn" href="/2019/08/23/第1篇：ATAC-seq的背景介绍以及与ChIP-Seq的异同/#more" rel="contents">
              阅读全文 &raquo;
            </a>
          </div>
          <!--/noindex-->
        
      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      

      

      
      
        <div class="post-eof"></div>
      
    </footer>
  </div>
  
  
  
  </article>


    
  </section>

  


          </div>
          


          

        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      

      <section class="site-overview-wrap sidebar-panel sidebar-panel-active">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <p class="site-author-name" itemprop="name">六六</p>
              <p class="site-description motion-element" itemprop="description">Seize the day! Collect the moment! Share the memory!</p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">7</span>
                  <span class="site-state-item-name">日志</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">2</span>
                  <span class="site-state-item-name">分类</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">2</span>
                  <span class="site-state-item-name">标签</span>
                </a>
              </div>
            

          </nav>

          
            <div class="feed-link motion-element">
              <a href="/atom.xml" rel="alternate">
                <i class="fa fa-rss"></i>
                RSS
              </a>
            </div>
          

          

          
          

          
          

          

        </div>
      </section>

      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">六六</span>

  
    <span class="post-meta-divider">|</span>
    <span class="post-meta-item-icon">
      <i class="fa fa-area-chart"></i>
    </span>
    
      <span class="post-meta-item-text">Site words total count&#58;</span>
    
    <span title="Site words total count">7.6k</span>
  
</div>


  <div class="powered-by">由 <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a> 强力驱动</div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">主题 &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Gemini</a> v5.1.4</div>




<div class="theme-info">
  <div class="powered-by"></div>
  <span class="post-count">博客全站共7.6k字</span>
</div>


        







        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  


  











  
  
    <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>
  

  
  
    <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>
  

  
  
    <script type="text/javascript" src="/lib/canvas-nest/canvas-nest.min.js"></script>
  


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.4"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.4"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.4"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.4"></script>



  

  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.4"></script>



  


  




	





  





  












  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  

  
  

  

  

  

</body>
</html>
